opencog/rule-engine/forwardchainer/ForwardChainer.cc

```diff
@@ -124,10 +125,19 @@ void ForwardChainer::do_step(void)
     _log->debug( "Derived rule size = %d", derived_rhandles.size());
 
     HandleSeq products;
-    //Applying all partial groundings found.
+    //Applying all partial/full groundings.
     for (Handle rhandle : derived_rhandles) {
-        HandleSeq hs = apply_rule(rhandle,_search_focus_Set);
-        products.insert(products.end(),hs.begin(),hs.end());
+        HandleSeq hs;
+        //Check for fully grounded outputs returned by derive_rules.
+        if (not contains_atomtype(rhandle, VARIABLE_NODE)) {
```

Comment:
I was going to suggest postponing the creation of the BindLink until apply_rule (since it doesn't look to be needed anywhere else anyway), but it seems it will take quite a bit of changes...